### PR TITLE
Coating protection

### PR DESCRIPTION
**Pull request by slacAWallace**
Implemented some fast faults to protect coatings on M1K4. Made a function block to do this, which I am sure we'll promptly throw away once the state-machine is working.

Also did a bit of formatting on the main program to improve readability. I was dying trying to scroll through it, so I indented things beneath comments and TC creates an expander/collapser so you can collapse a bunch of lines to reduce the length of code to scroll through. It mostly... works...

Also updated to the latest ads-ioc since it has some good diagnostic features.

### CODE REVIEW

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  <POU Name="P_StripeProtections" Id="{40ffbfa8-833e-45b2-a6cd-32aa1c1e46a6}" SpecialFunc="None">
+    <Declaration><![CDATA[PROGRAM P_StripeProtections
+
+(*
+MR1K4 (RBV = MR1K4:SOMS:ENC:YUP:CNT_RBV)
+B4C coating: RBV <= 15998960: 1111_1111_1111_1000 (allow everything above 400eV)
+Transition: 15998960 < RBV < 19998960: 0000_0000_0000_0000
+Silicon surface: RBV >= 19998960: 0000_0000_0000_0111 (allow everything below 400 eV)
+
+MR2K4 (RBV = MR2K4:KBO:MMS:Y.RBV)
```
Comment (sfsyunusSLAC) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:12]: any reason this doesn't use `MR2K4:KBO:ENC:Y:CNT_RBV`? Are we missing the encoder counts for these values?

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+B4C coating: RBV <= 4.7mm: 1111_1111_1111_1000 (allow everything above 400 eV)
+Transition: 4.7mm < RBV < 8.7mm: 0000_0000_0000_0000
+Silicon surface: RBV >= 8.7mm: 0000_0000_0000_0111 (allow everything below 400 eV)
+
+MR3K4 (RBV = MR3K4:KBO:MMS:X.RBV)
```
Comment (sfsyunusSLAC) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:17]: Same, as above `MR3K4:KBO:ENC:Y:CNT_RBV`

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    <Declaration><![CDATA[PROGRAM P_StripeProtections
+
+(*
+MR1K4 (RBV = MR1K4:SOMS:ENC:YUP:CNT_RBV)
+B4C coating: RBV <= 15998960: 1111_1111_1111_1000 (allow everything above 400eV)
+Transition: 15998960 < RBV < 19998960: 0000_0000_0000_0000
+Silicon surface: RBV >= 19998960: 0000_0000_0000_0111 (allow everything below 400 eV)
+
+MR2K4 (RBV = MR2K4:KBO:MMS:Y.RBV)
+B4C coating: RBV <= 4.7mm: 1111_1111_1111_1000 (allow everything above 400 eV)
```
Comment (sfsyunusSLAC) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:13]: Are these values from Matt Seaberg?

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU (hunk 1 above)
Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:12]: Yes. The encoder counts are being archived now so Matt will give us the ranges after the next time he switches them.

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU (hunk 3 above)
Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:13]: They are. I should credit him.

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 4]
```diff
@@ -0,0 +1,67 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.18">
+  <POU Name="P_StripeProtections" Id="{40ffbfa8-833e-45b2-a6cd-32aa1c1e46a6}" SpecialFunc="None">
+    <Declaration><![CDATA[PROGRAM P_StripeProtections
+VAR
+(*
+MR1K4 (RBV = MR1K4:SOMS:ENC:YUP:CNT_RBV)
+B4C coating: RBV <= 15998960: 1111_1111_1111_1000 (allow everything above 400eV)
+Transition: 15998960 < RBV < 19998960: 0000_0000_0000_0000
+Silicon surface: RBV >= 19998960: 0000_0000_0000_0111 (allow everything below 400 eV)
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: I think these are the old notes from before we changed the eV ranges. This should read:
`B4C coating: RBV <= 15998960: 1111_1111_1111_1000 (allow everything 350eV and 2.3keV)`  
`Transition: 15998960 < RBV < 19998960: 0000_0000_0000_0000`  
`Silicon surface: RBV >= 19998960: 0000_0000_0001_1111 (allow everything below 450 eV)`
Same is somewhat true for MR2K4 and MR3K4, but the code looks right to me in all cases.

Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: If it's between then I think maybe the left most bit should be 0 so for B4C it should be 0111_1111_1111_1000. The bitmask is super weird. It's below for all bits, except the most significant bit which enables everything above the corresponding threshold.

Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: Ok, let's change the B4C bitmask to 0111_1111_1111_1000 for all cases in that case.

Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: I might have made a mistake on silicon too. I guess at the moment everything below 250 eV is allowed, right? I think we should change silicon to 0000_0000_0001_1110.

Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: I'm sorry @mseaberg , I totally confused myself. I am writing this up in the confluence pages right now to make sure it's clear going forward. I think you had it right.

Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: Ok, sounds good. I'll wait to comment further until I read the page.

Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: https://confluence.slac.stanford.edu/display/L2SI/PMPS+Photon+Energy

Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: So if I understand correctly, B4C staying as is (1111_1111_1111_1000) will produce a fault at >2.3 keV. However, if I want to exclude silicon below 250 eV, I should change this to 0000_0000_0001_1110.

Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: Yes.

Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: Great.

Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]:
```suggestion
B4C coating: RBV <= 15998960: 1111_1111_1111_1000 (allow everything between 450eV and 2.3keV)
Transition: 15998960 < RBV < 19998960: 0000_0000_0000_0000
Silicon surface: RBV >= 19998960: 0000_0000_0001_1110 (allow everything between 250eV and 450eV)
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+From M. Seaberg
+*)		
+	fbStripProtMR1K4 : FB_MirrorTwoCoatingProtection := (
+		sDevName := 'MR1K4',
+		nUpperCoatingBoundary := 19998960,
+		nUpperCoatingBitmask := 2#0000_0000_0001_1111,
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:16]:
```suggestion
		nUpperCoatingBitmask := 2#0000_0000_0001_1110,
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+		sUpperCoatingType := 'Si',
+		nLowerCoatingBoundary := 15998960,
+		nLowerCoatingBitMask := 2#1111_1111_1111_1000,
+		sLowerCoatingType := 'B4C');
+(*
+MR2K4 (RBV = MR2K4:KBO:MMS:Y.RBV)
+B4C coating: RBV <= 4.7mm: 1111_1111_1111_1000 (allow everything above 400 eV)
+Transition: 4.7mm < RBV < 8.7mm: 0000_0000_0000_0000
+Silicon surface: RBV >= 8.7mm: 0000_0000_0001_1111 (allow everything below 400 eV)
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:25]:
```suggestion
MR2K4 (RBV = MR2K4:KBO:ENC:Y:CNT_RBV)
B4C coating: RBV <= 6976835: 1111_1111_1111_1000 (allow everything between 350eV and 2.3keV)
Transition: 6976835 < RBV < 7776781: 0000_0000_0000_0000
Silicon surface: RBV >= 7776781: 0000_0000_0001_1110 (allow everything below 250eV and 450eV)		
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+MR2K4 (RBV = MR2K4:KBO:MMS:Y.RBV)
+B4C coating: RBV <= 4.7mm: 1111_1111_1111_1000 (allow everything above 400 eV)
+Transition: 4.7mm < RBV < 8.7mm: 0000_0000_0000_0000
+Silicon surface: RBV >= 8.7mm: 0000_0000_0001_1111 (allow everything below 400 eV)		
+*)
+
+	fbStripProtMR2K4 : FB_MirrorTwoCoatingProtection := (
+		sDevName := 'MR2K4',
+		nUpperCoatingBoundary := 7776781,
+		nUpperCoatingBitmask := 2#0000_0000_0001_1111,
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:31]:
```suggestion
		nUpperCoatingBitmask := 2#0000_0000_0001_1110,
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+		sUpperCoatingType := 'Si',
+		nLowerCoatingBoundary := 6976835,
+		nLowerCoatingBitMask := 2#1111_1111_1111_1000,
+		sLowerCoatingType := 'B4C');
+(*
+MR3K4 (RBV = MR3K4:KBO:MMS:X.RBV)
+B4C coating: RBV >= -3mm: 1111_1111_1111_1000 (allow everything above 400 eV)
+Transition: -7mm < RBV < -3mm: 0000_0000_0000_0000
+Silicon surface: RBV <= -7mm: 0000_0000_0001_1111 (allow everything below 400 eV)
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:40]:
```suggestion
MR3K4 (RBV = MR3K4:KBO:ENC:X:CNT_RBV)
B4C coating: RBV >= 5620000: 1111_1111_1111_1000 (allow everything between 350eV and 2.3keV)
Transition: 4820000 < ENC < 5620000: 0000_0000_0000_0000
Silicon surface: RBV <= 4820000: 0000_0000_0001_1110 (allow everything between 250eV and 450 eV)
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+*)
+	fbStripProtMR3K4 : FB_MirrorTwoCoatingProtection := (
+		sDevName := 'MR3K4',
+		nUpperCoatingBoundary := 5620000,
+		nUpperCoatingBitmask := 2#1111_1111_1111_1000,
+		sUpperCoatingType := 'B4C',
+		nLowerCoatingBoundary := 4820000,
+		nLowerCoatingBitMask := 2#0000_0000_0001_1111,
```
Comment (mseaberg) [inline lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:48]:
```suggestion
		nLowerCoatingBitMask := 2#0000_0000_0001_1110,
```

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU (hunk 4 above)
Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU:10]: Ok, I went through and made suggestions for the small changes I think should happen based on this conversation. The only other file I looked at was tmo_optics.db to confirm PV's, I'm sure my review isn't helpful/needed on any of the other files.

Comment (slacAWallace) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU]: OK these changes have been pulled and will be deployed.

**Review by mseaberg — APPROVED** [review]

File: lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU (hunk 4 above)
Comment (mseaberg) [reply lcls-plc-tmo-optics/tmo_optics/POUs/P_StripeProtections.TcPOU]: Awesome, thanks.

**Pull request merged**